File: Day38/Part2/min_cost_make_array_equal.py
from typing import List

def min_cost(nums: List[int], cost: List[int]) -> int:
    A = sorted(zip(nums, cost))
    total_w = sum(cost)
    half = (total_w + 1) // 2
    acc = 0
    median = A[-1][0]
    for v, w in A:
        acc += w
        if acc >= half:
            median = v
            break
    res = 0
    for v, w in A:
        res += w * abs(v - median)
    return res

if __name__ == "__main__":
    print(min_cost([1,3,5,2], [2,3,1,14]))  # Example style

File: learnLeetCode_remote/Day38/Part2/min_cost_make_array_equal.cpp
#include <bits/stdc++.h>
using namespace std;

long long min_cost(vector<int> nums, vector<int> cost){
    vector<pair<long long,long long>> a; a.reserve(nums.size());
    for(size_t i=0;i<nums.size();++i) a.emplace_back(nums[i], cost[i]);
    sort(a.begin(), a.end());
    long long total_w=0; for(auto &p: a) total_w += p.second;
    long long half = (total_w + 1)/2;
    long long acc=0; long long median=a.back().first;
    for(auto &p: a){ acc += p.second; if(acc>=half){ median=p.first; break; } }
    long long res=0; for(auto &p: a) res += p.second * llabs(p.first - median);
    return res;
}

int main(){ ios::sync_with_stdio(false); cin.tie(nullptr);
    cout << min_cost(vector<int>{1,3,5,2}, vector<int>{2,3,1,14}) << "\n";
    return 0; }
